### PR TITLE
feat(gemm): Add mxfp8 gemm and quantize kernel

### PR DESCRIPTION
**Pull request by RuibinCheung**
* Add mxfp8 quantize and dequantize triton kernel.
* Refine quantize impl file structure.
* Support mxfp8 hipblaslt gemm kernel.
image
P.S: The prefix of `RR` (rocRoller) means this kernel is mxfp8 dtype.

### CODE REVIEW

File: csrc/pytorch/gemm/hipblaslt_gemm.cpp [hunk 1]
```diff
@@ -121,15 +121,23 @@ at::Tensor hipblaslt_gemm_fp8(at::Tensor A, at::Tensor scaleA_inv, at::Tensor B,
...
     PRIMUS_TURBO_CHECK(is_8bit_floating_point_dtype(B.scalar_type()));
     PRIMUS_TURBO_CHECK(is_16bit_floating_point_dtype(out_dtype));
-    PRIMUS_TURBO_CHECK(scaleA_inv.scalar_type() == at::kFloat);
-    PRIMUS_TURBO_CHECK(scaleB_inv.scalar_type() == at::kFloat);
+    if (granularity != "MX_BLOCKWISE") {
+        PRIMUS_TURBO_CHECK(scaleA_inv.scalar_type() == at::kFloat);
+        PRIMUS_TURBO_CHECK(scaleB_inv.scalar_type() == at::kFloat);
+    } else {
+        // scaling factor is e8m0 format.
+        PRIMUS_TURBO_CHECK(scaleA_inv.scalar_type() == at::kByte);
```
Comment (xiaobochen-amd) [inline csrc/pytorch/gemm/hipblaslt_gemm.cpp]: torch kFloat8_e8m0fnu

File: primus_turbo/pytorch/ops/quantization.py [hunk 2]
```diff
@@ -33,9 +38,14 @@ def quantize_fp8(
...
-        if axis is None:
-            raise ValueError("axis must be specified for rowwise FP8 quantization")
+        assert axis is not None, "The axis must be specified for rowwise FP8 quantization"
+
         return quantize_fp8_rowwise_impl(x, out_dtype, axis, scale)
+    elif granularity == ScalingGranularity.MX_BLOCKWISE:
+        assert block_size == MX_BLOCK_SIZE, f"The block size must be {MX_BLOCK_SIZE} for MXFP8 quantization"
+        assert scale is None, "The scale is not supported for MXFP8 quantization"
+
+        return quantize_mxfp8_impl(x, out_dtype, block_size)
```
Comment (xiaobochen-amd) [inline primus_turbo/pytorch/ops/quantization.py]: In the future, if more layouts are supported, quantize_mxfp8_impl can add an axis argument to specify along which dimension the 1×32 or 32x1 quantization is applied.

File: primus_turbo/pytorch/ops/quantization.py [hunk 3]
```diff
@@ -9,21 +9,26 @@
 import torch
 
 from primus_turbo.pytorch.core.float8 import ScalingGranularity
-from primus_turbo.pytorch.kernels.quantization_impl import (
+from primus_turbo.pytorch.kernels.quantize.quantize_impl import (
```
[the remark this answers is not in the document]
Comment (RuibinCheung) [reply primus_turbo/pytorch/ops/quantization.py]: ACK. Thx~

File: primus_turbo/pytorch/ops/quantization.py (hunk 2 above)
Comment (RuibinCheung) [reply primus_turbo/pytorch/ops/quantization.py]: ACK. Thx~

**Comment by RuibinCheung** [comment]
> How is the performance? Both for mx quantization and gemm.

The performance of mxfp8 gemm kernel is pretty bad. The throughput of fp8 GEMM is around 3000 TFLOP/s while the throughput of mxfp8 GEMM ranges from 600 TFLOP/s to 1150 TFLOP/s. I'm collecting gemm shapes to ask rocRoller team to tuning.

File: primus_turbo/pytorch/ops/gemm_fp8.py [hunk 4]
```diff
@@ -336,6 +337,129 @@ def backward(ctx, grad_out: torch.Tensor):
         return a_grad, b_grad, None, None, None, None
 
 
+class FP8GemmMXFunction(torch.autograd.Function):
+
+    MXFP8_GEMM_BACKEND = "hipblaslt"
+
+    @staticmethod
+    def get_fp8_dtype(format: Format, is_fwd_stage: bool):
+        if format == Format.E4M3:
```
[the remark this answers is not in the document]
Comment (RuibinCheung) [reply primus_turbo/pytorch/ops/gemm_fp8.py]: > It`s ok. we can include E5M2 as well. Megatron currently uses only E4M3, but adding E5M2 on our side shouldn’t be an issue. Would supporting E5M2 require a lot of additional code changes?

No more additional code change. I will add. Thanks.
